### PR TITLE
Fix handling of typecasting in `searchsorted`

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description

Fixes: #13902 

This PR fixes a type-casting issue with `searchsorted` where typecast was done to the `values` dtype instead of inspecting both input and values columns and converting them to common dtypes.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/frame.py [hunk 1]
```diff
@@ -1409,12 +1409,23 @@ def searchsorted(
...
-            for col, val in zip(self._columns, values)
+            if is_dtype_equal(col.dtype, common_dtype)
+            else col.astype(common_dtype)
+            for col, common_dtype in zip(self._columns, common_dtype_list)
         ]
+        values = [
+            val
+            if is_dtype_equal(val.dtype, common_dtype)
+            else val.astype(common_dtype)
+            for val, common_dtype in zip(values, common_dtype_list)
```
Comment (wence-) [inline python/cudf/cudf/core/frame.py:1426]: This allows:
```
s = cudf.Series([1, 2, 3], index=[1, 2, 3])
s.loc["1":"3"]  # same as if we wrote s.loc[1:3]
```
to succeed, even though `s.loc["1"]` and `s.loc["3"]` raise `KeyError`. But I note that pandas does the same thing here, so perhaps this is safe (@mroeschke is this a bug?). If I make a series:
```
import pandas as pd
s = pd.Series([1, 2, "1"])
s.sort_values() # raise ValueError
# But
s = pd.Series([1, 2])
s.searchsorted("1") # => array([0]), whereas I was expecting ValueError
```

Comment (mroeschke) [reply python/cudf/cudf/core/frame.py:1426]: Interesting. I think we are mirroring this behavior from numpy but I think pandas should be raising a ValueError here too

```python
In [4]: import numpy as np

In [5]: arr = np.array([1, 2])

In [6]: arr.searchsorted("1")
Out[6]: 0
```

Comment (galipremsagar) [reply python/cudf/cudf/core/frame.py:1426]: @mroeschke Would it be possible for you to open a pandas issue to track this? I can link it in our code for future change/reference.

Comment (mroeschke) [reply python/cudf/cudf/core/frame.py:1426]: Sure, https://github.com/pandas-dev/pandas/issues/54668

Comment (galipremsagar) [reply python/cudf/cudf/core/frame.py:1428]: Thanks @mroeschke ! 

@wence- I added a TODO to update our code behavior in future based on how https://github.com/pandas-dev/pandas/issues/54668 progresses. Let me know if you think that's okay.

File: python/cudf/cudf/tests/test_indexing.py [hunk 2]
```diff
@@ -2067,3 +2067,16 @@ def test_loc_index_notinindex_slice(
             expect = pdf.loc[lo:hi:take_order]
             actual = df.loc[lo:hi:take_order]
             assert_eq(expect, actual)
+
```
Comment (wence-) [inline python/cudf/cudf/tests/test_indexing.py:2070]: If pandas upgrades are an intentional thing, then we can encode the pandas behaviour, and hence the TODO, in a test:

```
@pytest.mark.xfail(reason="https://github.com/pandas-dev/pandas/issues/54668")
def test_searchsorted_mixed_str_int():
    s = pd.Series([1, 2, 3], dtype="int")
    c = cudf.from_pandas(s)
    with pytest.raises(ValueError):
        actual = s.searchsorted("a")
    with pytest.raises(ValueError):
        expect = c.searchsorted("a")
    assert_eq(expect, actual)
```

This test will fail once that pandas issue is addressed, and remind us (when upgrading pandas) to change the cudf behaviour.

**Review by wence- — APPROVED** [review]
Minor suggestion for a test, but looks good.

File: python/cudf/cudf/tests/test_indexing.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_indexing.py:2070]: Done 👍

**Pull request merged**